[azure-service-bus] Add possibility to receive messages from Azur… (#4876)
vividus-framework/vividus: docs/modules/plugins/pages/plugin-azure-service-bus.adoc

[source,gherkin]
----
When I send message to `$serviceBusKey` service bus with payload:`$payload`
----

* `$serviceBusKey` - The key associated with the Azure Service Bus client, will be used to find required properties values.
* `$payload` - Message to send to the service bus.

.Send simple message to the Service Bus Queue
[source,gherkin]
----
When I send message to `myProjectQueue` service bus with payload:`Hello World!`
----

=== Start consuming messages

Starts the Azure Service Bus consumer with the provided configuration to listen the specified
https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview#concepts[Queue or Topic].

IMPORTANT: The consumer must be stopped when it's not needed.

[source,gherkin]
----
When I start consuming messages from `$serviceBusKey` service bus
----

* `$serviceBusKey` - The key associated with the Azure Service Bus client, will be used to find required properties values.

.Start consuming messages from the topic configured by Service Bus key `myProjectTopic`
[source,gherkin]
----
When I start consuming messages from `myProjectTopic` service bus
----

> === Stop consume messages

Reviewer: ```suggestion
=== Stop consuming messages
```

[source,gherkin]
----
When I wait with `$timeout` timeout until count of consumed by `$serviceBusKey` service bus messages is $comparisonRule `$expectedCount`
----

* `$timeout` - The maximum time to wait for the events in {durations-format-link} format.
* `$serviceBusKey` - The key associated with the Azure Service Bus client, will be used to find required properties values.
* `$comparisonRule` - xref:parameters:comparison-rule.adoc[The comparison rule].
* `$expectedCount` - The expected count of the events to be matched by the rule.

.Wait for at least 5 messages from the queue for 30 seconds
[source,gherkin]
----
When I wait with `PT30S` timeout until count of consumed by `myProjectQueue` service bus messages is >= `5`
----

=== Save messages to the variable

Drains/Peeks the consumed messages to the specified variable. If the consumer is not stopped, the new events might arrive after the draining.
If the consumer is stopped, all the events received from the consumer start or after the last draining operation are stored to the variable.

[source,gherkin]
----

> When I $queueOperation consumed `$serviceBusKey` service bus messages to $scopes variable `$variableName`

Reviewer: I don't think it's for queue only

----

* `$queueOperation` - `DRAIN` - saves the messages consumed since the last drain or from the consumption start and moves the consumer cursor to the position after the last consumed message;

> `PEEK` - saves the messages consumed since the last drain or from the consumption start and doesn't change the consumer cursor position.

Reviewer 2: does it makes sense to represent operations as sublist of queueOperation?

* `$serviceBusKey` - The key associated with the Azure Service Bus client, will be used to find required properties values.
* `$scopes` - xref:commons:variables.adoc#_scopes[The comma-separated set of the variables scopes].
* `$variableName` - The variable name to store the messages. The messages are accessible via zero-based index, e.g. `${my-var[0]}` will return the first received message.

.Peek messages from the consumer and save to variable
[source,gherkin]
----

> When I PEEK consumed `myProjectQueue` service bus messages to STORY variable `queueMessages`

Reviewer: please add a set of complete examples to get understanding how to use these steps (in the same way we have for Kafka)

include::partial$azure-profile-and-subscription.adoc[]

=== Azure Service Bus Configuration

Azure Service Bus configurations are established via set of properties with the following format:

[source,properties]
----
azure.service-bus.{service-bus-key}.{property-name}=property value
----

where:

. `service-bus-key` - The key associated with the Azure Service Bus configuration, which will be used as a step parameter.

> . `property-name` - The name of the topic property. One of:

Reviewer: ```suggestion
. `property-name` - The name of Service Bus property. One of:
```

> .. `channel-type` - The type of service bus messaging components: either

Reviewer: ```suggestion
.. `channel-type` - The type of Service Bus messaging channel: either
```

https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview#concepts[QUEUE or TOPIC].

> .. `namespace` - The name of the namespace the service bus belongs to.

Reviewer: ```suggestion
.. `namespace` - The name of the namespace Service Bus belongs to.
```

.. `name` - The queue or topic name.

> .. `subscription-name` - The name of the topic subscription. Only for TOPIC channel-type.

Reviewer: ```suggestion
.. `subscription-name` - The name of the topic subscription. Only for `TOPIC` channel type.
```

== Steps

=== Send message

Send message to the specified Azure Service Bus https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview#concepts[Queue or Topic].

[source,gherkin]
----
When I send message to `$serviceBusKey` service bus with payload:`$payload`
----

* `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.
* `$payload` - Message to send to the service bus.

.Send simple message to the Service Bus Queue
[source,gherkin]
----
When I send message to `myProjectQueue` service bus with payload:`Hello World!`
----

=== Start consuming messages

> Starts the Azure Service Bus consumer with the provided configuration to listen the specified

Reviewer: ```suggestion
Starts Azure Service Bus consumer with the provided configuration to listen the specified
```

https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-messaging-overview#concepts[Queue or Topic].

> IMPORTANT: The consumer must be stopped when it's not needed.

Reviewer: ```suggestion
IMPORTANT: The consumer must be stopped when it's not used.
```

[source,gherkin]
----
When I start consuming messages from `$serviceBusKey` service bus
----

* `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.

.Start consuming messages from the topic configured by Service Bus key `myProjectTopic`
[source,gherkin]
----
When I start consuming messages from `myProjectTopic` service bus
----

=== Stop consuming messages

> Stops the Azure Service Bus consumer started by the <<_start_consuming_messages>> step before.

Reviewer: ```suggestion
Stops Azure Service Bus consumer started by the <<_start_consuming_messages>> step before.
```

All recorded messages are kept and can be drained into the variable using the step <<_save_messages_to_the_variable>>.

[source,gherkin]
----
When I stop consuming messages from `$serviceBusKey` service bus
----

* `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.

.Stop consuming messages from the topic configured by Service Bus key `myProjectTopic`
[source,gherkin]
----
When I stop consuming messages from `myProjectTopic` service bus
----

=== Wait for the messages

Waits until the count of the consumed messages (from the consumer start or after the last draining operation) matches to the rule or until the timeout is exceeded.

[source,gherkin]
----
When I wait with `$timeout` timeout until count of consumed by `$serviceBusKey` service bus messages is $comparisonRule `$expectedCount`
----

* `$timeout` - The maximum time to wait for the events in {durations-format-link} format.

> * `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.

Reviewer: ```suggestion
* `$serviceBusKey` - The key associated with Azure Service Bus configuration, will be used to find required properties values.
```

* `$comparisonRule` - xref:parameters:comparison-rule.adoc[The comparison rule].
* `$expectedCount` - The expected count of the events to be matched by the rule.

.Wait for at least 5 messages from the queue for 30 seconds
[source,gherkin]
----

> When I wait with `PT30S` timeout until count of consumed by `myProjectQueue` service bus messages is >= `5`

Reviewer: ```suggestion
When I wait with `PT30S` timeout until count of consumed `myProjectQueue` service bus messages is >= `5`
```

----

=== Save messages to the variable

Drains/Peeks the consumed messages to the specified variable. If the consumer is not stopped, the new events might arrive after the draining.
If the consumer is stopped, all the events received from the consumer start or after the last draining operation are stored to the variable.

[source,gherkin]
----
When I $operation consumed `$serviceBusKey` service bus messages to $scopes variable `$variableName`
----

* `$operation` - `DRAIN` - saves the messages consumed since the last drain or from the consumption start and moves the consumer cursor to the position after the last consumed message;
`PEEK` - saves the messages consumed since the last drain or from the consumption start and doesn't change the consumer cursor position.

> * `$serviceBusKey` - The key associated with the Azure Service Bus configuration, will be used to find required properties values.

Reviewer: ```suggestion
* `$serviceBusKey` - The key associated with Azure Service Bus configuration, will be used to find required properties values.
```

* `$scopes` - xref:commons:variables.adoc#_scopes[The comma-separated set of the variables scopes].
* `$variableName` - The variable name to store the messages. The messages are accessible via zero-based index, e.g. `${my-var[0]}` will return the first received message.

> 

Reviewer: == Examples
